sumologic-app-utils/src/awsresource.py

```diff
@@ -0,0 +1,1116 @@
+        except ClientError as e:
+            print("Error in updating trail %s" % e.response['Error'])
+            raise
+        except Exception as e:
+            print("Error in updating trail %s" % e)
+            raise
+
+    def delete(self, trail_name, *args, **kwargs):
+        try:
+            self.cloudtrailcli.delete_trail(
+                Name=trail_name
+            )
+            print("Trail deleted %s" % trail_name)
+        except ClientError as e:
+            print("Error in deleting trail %s" % e.response['Error'])
+            raise
+        except Exception as e:
+            print("Error in deleting trail %s" % e)
+            raise
+
+    def _transform_bool_values(self, k, v):
+        if k in self.boolean_params:
+            return True if v and v == "true" else False
+        else:
+            return v
+
+    def extract_params(self, event):
+        props = event.get("ResourceProperties")
+        parameters = ["S3BucketName", "S3KeyPrefix", "IncludeGlobalServiceEvents", "IsMultiRegionTrail",
+                      "EnableLogFileValidation", "IsOrganizationTrail"]
+        params = {k: self._transform_bool_values(k, v) for k, v in props.items() if k in parameters}
+        params['Name'] = props.get("TrailName")
+        return {
+            "props": props,
+            "trail_name": props.get("TrailName"),
+            "params": params
+        }
+
+
+class TagAWSResources(AWSResource):
+
+    def __init__(self, props, *args, **kwargs):
+        print('Tagging aws resource %s' % props.get("AWSResource"))
+
+    def _tag_aws_resources(self, region_value, aws_resource, tags, account_id, delete_flag, filter_regex):
+        # Get the class instance based on AWS Resource
+        tag_resource = AWSResourcesProvider.get_provider(aws_resource, region_value, account_id)
+
+        # Fetch and Filter the Resources.
+        resources = tag_resource.fetch_resources()
+        filtered_resources = tag_resource.filter_resources(filter_regex, resources)
+
+        if filtered_resources:
+            # Get the ARNs for all resources
+            arns = tag_resource.get_arn_list(filtered_resources)
+
+            # Tag or un-tag the resources.
+            if delete_flag:
+                tag_resource.delete_tags(arns, tags)
+            else:
+                tag_resource.add_tags(arns, tags)
+
+    def create(self, region_value, aws_resource, tags, account_id, filter_regex, *args, **kwargs):
+        print("TAG AWS RESOURCES - Starting the AWS resources Tag addition with Tags %s." % tags)
+        regions = [region_value]
+        for region in regions:
+            self._tag_aws_resources(region, aws_resource, tags, account_id, False, filter_regex)
+        print("TAG AWS RESOURCES - Completed the AWS resources Tag addition.")
+
+        return {"TAG_CREATION": "Successful"}, "Tag"
+
+    def update(self, region_value, aws_resource, tags, account_id, filter_regex, *args, **kwargs):
```

Comment:
It is documented in AWS Observability docs. We don't have a functionality doc for Sumo App utils.

Reply:
and customers are ok if the old tags remain as it is?

Reply:
We do not ask for Tags from Customers. We just take the value. So, we do not change the Key for tags just the value.

But yes, in general it can be improved, if we keep track of old tag keys that were passed.

Reply:
cool then we should add this to improvement jira.